### PR TITLE
Update ticktick extension quick add with date/list/priority syntax

### PR DESCRIPTION
**Pull request by Vernal**
## Description

Enhances TickTick's **Quick Add Task** command with a practical subset of the syntax supported by TickTick's web and desktop apps.

- Parses natural-language dates and times, including time-only input that defaults to today
- Supports explicit `*date` syntax, `~List`/`^List` selection, and `!priority`
- Creates all-day tasks for date-only input and timed tasks when a time is present
- Removes recognized metadata from the saved task title
- Rejects metadata-only input instead of creating a blank or misleading task
- Leaves ambiguous numeric ranges such as `5-7` in the title unless explicitly marked as `*5-7`
- Handles repeated metadata safely and preserves unrelated title text

Examples:

- `Call the client at 9am` → due today at 9:00 AM
- `Review the report next Monday` → all-day task due next Monday
- `Submit expenses *2026-08-25 ~Work !high` → high-priority task in Work due August 25

The parser intentionally does not handle tags, recurrence, or assignees because those fields are not exposed by TickTick's macOS AppleScript task-creation interface.

## Validation

- Added 11 parser test groups covering intended syntax and adversarial edge cases
- Ran `npm test`
- Ran `npm run build`
- Ran `npm run lint`
- Manually tested Quick Add in Raycast with time-only and date/time input

## Screencast

Not included because this is a no-view command and the change affects the resulting TickTick task rather than Raycast UI.

## Checklist

- [x] I read the [extension guidelines](https://developers.raycast.com/basics/prepare-an-extension-for-store)
- [x] I read the [documentation about publishing](https://developers.raycast.com/basics/publish-an-extension)
- [x] I ran `npm run build` and tested the distribution build in Raycast
- [x] I checked that files in the `assets` folder are used by the extension itself
- [x] I checked that assets used by the README are placed outside of the `metadata` folder

### CODE REVIEW

**Review by 0xdhrv — COMMENTED** [review]
Hi Vernal 👋

Thanks for your contribution 🔥

@viduycheung do you want to check this?

File: extensions/ticktick/src/utils/quickAddParser.ts [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+    token === "!none" || token === "!0"
+      ? "0"
+      : token === "!low" || token === "!3" || token === "!"
+      ? "1"
+      : token === "!medium" || token === "!med" || token === "!2" || token === "!!"
+      ? "3"
+      : "5";
+
+  return { text: `${text.slice(0, match.index)} ${text.slice(match.index + match[0].length)}`, priority };
+};
```
Comment (viduycheung) [inline extensions/ticktick/src/utils/quickAddParser.ts:36]: The priority parsing should be adjusted to stay consistent with TickTick’s in-app Quick Add behavior.

I think we should only support:

!1 / !high   -> high
!2 / !medium -> medium
!3 / !low    -> low
!none         -> none

The additional aliases such as !, !!, !!!, !0, and !med should be removed unless they are also supported by TickTick itself.

Keeping the syntax aligned with TickTick avoids introducing a separate set of Quick Add rules in the Raycast extension and also reduces the risk of accidentally parsing characters like ! from a normal task title as priority metadata.

It would also be good to add test cases for each supported priority syntax to make sure the mappings stay consistent with TickTick.

**Review by viduycheung — COMMENTED** [review]
For multiple metadata values, I think the parser should avoid silently consuming tokens that are not actually used.

For example:

```text
Review report ~Work ~Personal
Fix bug !low !high
Prepare Monday report for Friday
```

The current behavior can keep the first parsed value while still removing the later matching tokens from the task title. That makes the parsing lossy: some user input disappears even though it has no effect on the final task.

I think we should define a clear strategy here, for example:

* first match wins, and only that match is removed from the title; or
* last match wins, if that better matches TickTick’s behavior.

But `first match wins + remove all matching tokens` is surprising and should be avoided.

The general rule should be: **only remove metadata that is actually used in the final parsed result.**

For unintended parsing, I think the parser should also be more conservative.

Since date expressions and other metadata can appear naturally in a task title, valid titles such as:

```text
Monday report
Friday deployment checklist
Prepare Monday report for Friday
Move Tuesday meeting to Thursday
```

may be interpreted as metadata even when the user intended them to remain part of the title.

This is especially problematic because parsing is destructive: once a token is recognized, it is removed from the title.

For Quick Add, I think false positives are more harmful than false negatives. If something is not parsed, the user still gets the original text in the task title. If something is incorrectly parsed, part of the user’s input is silently changed or lost.

So I would prefer a more conservative parsing strategy, with tests covering ambiguous cases where date/project/priority-like text is intended to be part of the task title.

**Comment by Vernal** [comment]
Thanks for the detailed feedback. I agreed that the existing first-match-wins/remove-all behavior was lossy and have updated the implementation:

- Projects now use the first recognized project and remove only that token; later project-like tokens remain in the title.
- Dates use the first recognized date and remove only that expression; later date expressions remain in the title.
- Priority syntax is now limited to TickTick’s supported forms: `!1`/`!high`, `!2`/`!medium`, `!3`/`!low`, and `!none`. The undocumented aliases have been removed.
- Multiple priorities use the last recognized value, matching the behavior observed in TickTick.
- Tests now cover every supported priority form and the ambiguous examples from the review.

I also tested the date examples directly in the current TickTick web app. TickTick produced the following behavior:

- `Monday report` → title `report`, due Monday
- `Friday deployment checklist` → title `deployment checklist`, due Friday
- `Prepare Monday report for Friday` → Monday is used; Friday remains in the title
- `Move Tuesday meeting to Thursday` → Tuesday is used; Thursday remains in the title
- `today tomorrow` → due today, title `tomorrow`

I also tested `Fix bug !low !high`; TickTick created `Fix bug` with high priority, so its priority behavior is last-match-wins while consuming both recognized priority tokens.

Because this PR is intended to make Raycast Quick Add behave like TickTick’s native Quick Add, I retained TickTick’s bare-weekday recognition and added parity tests for it. Would you be comfortable retaining that native behavior, or would you prefer the extension to intentionally be more conservative than TickTick itself?

File: extensions/ticktick/src/utils/quickAddParser.ts [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+
+  return { text, isAllDay: false };
+};
+
+/** Parse the subset of TickTick quick-add syntax supported by the macOS AppleScript API. */
+export const parseQuickAdd = (input: string, projects: Project[], now = new Date()): ParsedQuickAdd => {
+  let remainingText = input;
+  const project = parseProject(remainingText, projects);
+  const projectId = project.projectId;
+  remainingText = project.text;
```
[the remark this answers is not in the document]
Comment (Vernal) [reply extensions/ticktick/src/utils/quickAddParser.ts:219]: This is an expected outcome and, while not ideal in all situations, mirrors TickTick web app behavior.

**Comment by viduycheung** [comment]
Hi @Vernal 

I understand the motivation for keeping the parsing behavior consistent with TickTick, and I agree that the NLP parsing itself can be useful.

However, I think there is an important UX difference between TickTick and Raycast here.

In TickTick, the recognized parts are explicitly surfaced to the user, and the user has an opportunity to remove or cancel the detected metadata before creating the task. In Raycast, this kind of inline feedback/cancellation may not be possible in the same way.

Because of that, I think the current parsing is quite aggressive. If a user types something like:

```text
Monday report
```

and the parser automatically turns it into:

```text
title: report
due date: Monday
```

the user may not even realize that `Monday` was interpreted as metadata. If they didn't intend that, part of their original input is effectively lost.

I think this is especially important for NLP parsing because false positives are more problematic than false negatives. If we don't recognize something, the user still gets their original text. If we recognize something incorrectly, we can silently change the meaning of the task or remove part of the user's input.

One possible approach would be to introduce a separate command specifically for NLP-based task creation. That command could have a UI designed around the fact that parsing is involved, rather than making the existing Quick Add command silently interpret arbitrary natural language.

An even better UX could be to show the parsed result before creating the task and give the user an opportunity to confirm or edit what was detected. For example:

```text
Input:
Monday report

Detected:
Title: report
Due date: Monday

[Create] [Edit / Cancel]
```

This would make the NLP behavior explicit and avoid silently modifying the user's input.

So I'm not against supporting NLP parsing — I actually think it could be a useful feature. My concern is mainly that the current command doesn't give the user enough visibility or control once the parser decides to interpret part of their input as metadata.

**Comment by Vernal** [comment]
Thanks - I agree that recognized metadata should be visible before Quick Add modifies the task.
I’ve updated the command so that unmarked natural-language dates and times now show a confirmation dialog titled “Create This Task?”. It previews the parsed title, due date/time, list, and priority before offering Create Task or Cancel. Cancelling returns the user to Raycast, where they can edit the original input.

The existing Add Task command remains available when a full editable form is preferable. Explicit date/time syntax prefixed with * skips confirmation, preserving a genuinely fast path when parsing is intentional.

I also rechecked the date behavior against TickTick and corrected several details:
- A bare weekday matching today means today.
- next Thursday means Thursday of the following calendar week.
- next week means seven days from today.
- tonight means today at 8:00 PM.

The parser now has 17 passing test groups, and I reran lint, TypeScript validation, the production build, and the Quick Add cancellation flow locally.

**Comment by viduycheung** [comment]
I tried the latest changes, and I think the parsing itself is working much better now. However, after experiencing the new flow, I still have some concerns about changing the existing Quick Add behavior.

My preference would be to introduce a **separate command for NLP-based task creation**, rather than changing the existing Quick Add command.

The current Quick Add flow is already a very fast and predictable workflow. Adding automatic NLP parsing to it changes its behavior for existing users, especially when a word in the task title happens to look like a date or other metadata.

A separate command could make the distinction much clearer:

* **Quick Add** → keep the existing behavior and remain a fast, predictable input flow.
* **NLP Quick Add** (or a similar command) → explicitly opt into NLP parsing and the additional confirmation flow.

This way, users who want NLP parsing can choose the new command, while existing users don't have to worry about their input being unexpectedly interpreted.

I also tried the new confirmation UI. I think showing the parsed result is a good improvement, but currently the user can only choose whether to create the task or cancel the entire operation.

So my main suggestion is:

1. Keep the existing Quick Add command unchanged.
2. Add a separate command specifically for users who want NLP parsing.
3. In the NLP command, make the detected metadata editable or allow the user to opt out of individual detections before creating the task.

I think this would preserve the speed of the existing Quick Add flow while still providing a much richer NLP experience for users who explicitly want it.

**Comment by 0xdhrv** [comment]
Hey @Vernal 👋 

Thanks for your contribution!

Moving to draft until the changes are made as requested by @viduycheung 

Feel free to mark the PR as ready again and we will review it

**Comment by viduycheung** [comment]
@Vernal We’re planning to improve the NLP parsing through a macOS app using AppleScript, which should give us more flexibility to make the recognition more accurate and better integrated with the existing workflow.

Our plan is to eventually add a setting/toggle to the existing Quick Add command, so users can choose whether they want to enable NLP parsing. This way, the default Quick Add flow can remain unchanged, while users who want the NLP functionality can explicitly turn it on.

Given this direction, I think we can leave the changes we discussed earlier as they are for now. There’s no need to introduce a separate command or change the confirmation flow at this stage.

Once we have the improved NLP implementation through the macOS extension, we can revisit the UX and decide whether additional confirmation or editing of the detected metadata is still necessary.

Thanks for taking the time to iterate on this!

**Comment by Vernal** [comment]
Thanks for the update, that all makes sense! Glad to hear the NLP
parsing will get a proper upgrade down the line via the macOS app, no
concerns from me about that direction.

Since you said the current changes can stay as they are, I'll go ahead
and mark this ready for review again. Let me know if anything else comes
up.

**Comment by viduycheung** [comment]
@Vernal I think there’s a small misunderstanding 😄

We’ll wait for the macOS update and then handle the NLP adaptation ourselves. We’re not asking you to continue working on the NLP part of this PR.

So for now, you can simply close this PR. Thanks again for all the work and for addressing the feedback!

**Comment by Vernal** [comment]
Oh! No problem. Ok, to the future. Keep up the good work.
